src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractAddItemCommandHandler.cs

```diff
@@ -0,0 +1,122 @@
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IPhysicalProjectTree _projectTree;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IVsUIService<IVsAddProjectItemDlg> _addItemDialog;
+        private readonly IVsUIService<IVsShell> _vsShell;
+        private readonly Lazy<Dictionary<long, List<TemplateDetails>>> _commandMap;
+
+        [ImportingConstructor]
+        public AbstractAddItemCommandHandler(ConfiguredProject configuredProject, IPhysicalProjectTree projectTree, IUnconfiguredProjectVsServices projectVsServices, IVsUIService<IVsAddProjectItemDlg> addItemDialog, IVsUIService<SVsShell, IVsShell> vsShell)
+        {
+            _configuredProject = configuredProject;
+            _projectTree = projectTree;
+            _projectVsServices = projectVsServices;
+            _addItemDialog = addItemDialog;
+            _vsShell = vsShell;
+
+            _commandMap = new Lazy<Dictionary<long, List<TemplateDetails>>>(() => GetTemplateDetails());
+        }
+
+        protected abstract Dictionary<long, List<TemplateDetails>> GetTemplateDetails();
+
+        public Task<CommandStatusResult> GetCommandStatusAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, string commandText, CommandStatus progressiveStatus)
+        {
+            Requires.NotNull(nodes, nameof(nodes));
+
+            if (nodes.Count == 1 && _projectTree.NodeCanHaveAdditions(nodes.First()) && TryGetCommandDetails(commandId, out _))
+            {
+                return GetCommandStatusResult.Handled(commandText, CommandStatus.Enabled);
+            }
+
+            return GetCommandStatusResult.Unhandled;
+        }
+
+        public async Task<bool> TryHandleCommandAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut)
+        {
+            Requires.NotNull(nodes, nameof(nodes));
+
+            if (nodes.Count == 1)
+            {
+                if (TryGetCommandDetails(commandId, out TemplateDetails result))
+                {
+                    IProjectTree node = nodes.First();
+
+                    if (!_projectTree.NodeCanHaveAdditions(node))
+                    {
+                        return false;
+                    }
+
+                    __VSADDITEMFLAGS uiFlags = __VSADDITEMFLAGS.VSADDITEM_AddNewItems |
+                                               __VSADDITEMFLAGS.VSADDITEM_SuggestTemplateName |
+                                               __VSADDITEMFLAGS.VSADDITEM_AllowHiddenTreeView;
+
+                    string strBrowseLocations = _projectTree.TreeProvider.GetAddNewItemDirectory(node);
+                    string strFilter = string.Empty;
+                    Guid addItemTemplateGuid = Guid.Empty;  // Let the dialog ask the hierarchy itself
```

Comment:
I find it easier to parse if these declarations are closer to where they are actually used, ie below the LoadPackageString.